api/api.go

```diff
@@ -193,75 +184,71 @@ func (b *BlockChainAPI) GetTransactionByBlockHashAndIndex(
 	ctx context.Context,
 	blockHash common.Hash,
 	index hexutil.Uint,
-) *RPCTransaction { // todo add error to the return type
+) (*RPCTransaction, error) { // todo add error to the return type
 	block, err := b.blocks.GetByID(blockHash)
 	if err != nil {
-		b.logger.Error().Err(err).Msg("failed to get block by id")
-		return nil
+		return handleError[*RPCTransaction](b.logger, err)
 	}
 
 	highestIndex := len(block.TransactionHashes) - 1
 	if index > hexutil.Uint(highestIndex) {
-		return nil
+		return nil, nil
 	}
 
 	txHash := block.TransactionHashes[index]
 	tx, err := b.GetTransactionByHash(ctx, txHash)
 	if err != nil {
-		b.logger.Error().Err(err).Msg("failed to get transaction by id")
-		return nil
+		return handleError[*RPCTransaction](b.logger, err)
 	}
 
-	return tx
+	return tx, nil
 }
 
 // GetTransactionByBlockNumberAndIndex returns the transaction for the given block number and index.
 func (b *BlockChainAPI) GetTransactionByBlockNumberAndIndex(
 	ctx context.Context,
 	blockNumber rpc.BlockNumber,
 	index hexutil.Uint,
-) *RPCTransaction { // todo add error to the return type
+) (*RPCTransaction, error) { // todo add error to the return type
 	block, err := b.blocks.GetByHeight(uint64(blockNumber))
 	if err != nil {
-		if !errors.Is(err, storageErrs.NotFound) {
-			b.logger.Error().Err(err).Msg("failed to get block by height")
-		}
-		return nil
+		return handleError[*RPCTransaction](b.logger, err)
 	}
 
 	highestIndex := len(block.TransactionHashes) - 1
 	if index > hexutil.Uint(highestIndex) {
-		return nil
+		return nil, nil
 	}
 
 	txHash := block.TransactionHashes[index]
 	tx, err := b.GetTransactionByHash(ctx, txHash)
 	if err != nil {
-		if !errors.Is(err, storageErrs.NotFound) {
-			b.logger.Error().Err(err).Msg("failed to get transaction by hash")
-		}
-
-		return nil
+		return handleError[*RPCTransaction](b.logger, err)
 	}
 
-	return tx
+	return tx, nil
 }
 
 // GetTransactionReceipt returns the transaction receipt for the given transaction hash.
 func (b *BlockChainAPI) GetTransactionReceipt(
 	ctx context.Context,
 	hash common.Hash,
 ) (*types.Receipt, error) {
-	rcp, err := b.receipts.GetByTransactionID(hash)
+	_, err := b.transactions.Get(hash)
 	if err != nil {
-		if !errors.Is(err, storageErrs.NotFound) {
-			b.logger.Error().Err(err).Msg("failed to get transaction by ID")
-			return nil, errs.ErrInternal
-		}
+		return handleError[*types.Receipt](b.logger, err)
+	}
 
-		return nil, fmt.Errorf("failed to get transaction receipt: %w", err)
+	rcp, err := b.receipts.GetByTransactionID(hash)
+	if err != nil {
+		return handleError[*types.Receipt](b.logger, err)
 	}
 
+	// todo workaround until new version of flow-go is released
```

Comment:
Nice one :+1: I think the `evm.TransactionExecuted` event payload is going to contain `BlockHash`, so we can avoid the extra DB call.